Rubberduck.Parsing/Rewriter/RewriteSessionBase.cs

```diff
@@ -0,0 +1,90 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using NLog;
+using Rubberduck.Parsing.VBA.Parsing;
+using Rubberduck.VBEditor;
+
+namespace Rubberduck.Parsing.Rewriter
+{
+    public abstract class RewriteSessionBase : IRewriteSession
+    {
+        protected readonly IDictionary<QualifiedModuleName, IExecutableModuleRewriter> CheckedOutModuleRewriters = new Dictionary<QualifiedModuleName, IExecutableModuleRewriter>();
+        protected readonly IRewriterProvider RewriterProvider; 
+
+        private readonly Func<IRewriteSession, bool> _rewritingAllowed;
+
+        protected readonly Logger Logger = LogManager.GetCurrentClassLogger();
+        private readonly object _invalidationLockObject = new object();
+
+        public abstract CodeKind TargetCodeKind { get; }
+
+        protected RewriteSessionBase(IRewriterProvider rewriterProvider, Func<IRewriteSession, bool> rewritingAllowed)
+        {
+            RewriterProvider = rewriterProvider;
+            _rewritingAllowed = rewritingAllowed;
+        }
+
+
+        public IModuleRewriter CheckOutModuleRewriter(QualifiedModuleName module)
+        {
+            if (CheckedOutModuleRewriters.TryGetValue(module, out var rewriter))
+            {
+                return rewriter;
+            }
+            
+            rewriter = ModuleRewriter(module);
+            CheckedOutModuleRewriters.Add(module, rewriter);
+            return rewriter;
+        }
+
+        protected abstract IExecutableModuleRewriter ModuleRewriter(QualifiedModuleName module);
+
+        public bool TryRewrite()
+        {
+            if (!CheckedOutModuleRewriters.Any())
+            {
+                return false;
+            }
+
+            lock (_invalidationLockObject)
+            {
+                if (_isInvalidated)
```

Comment:
could we use the property accessor here instead? Is there any way for an invalidated session to get into a non-invalidated state?

Reply:
There is no way back. So, it is probably safe to use the property accessor.